src/saltext/vmware/utils/common.py

```diff
@@ -604,6 +604,60 @@ def get_path(node, service_instance, path=""):
     return get_path(node, service_instance, path)
 
 
+def get_clusters(service_instance, datacenter_name=None, cluster_name=None):
+    """
+    Returns clusters in a vCenter.
+
+    service_instance
+        The Service Instance Object from which to obtain cluster.
+
+    datacenter_name
+        (Optional) Datacenter name to filter by.
+
+    cluster_name
+        (Optional) Exact cluster name to filter by. Requires datacenter_name.
+    """
+    if datacenter_name is None and cluster_name:
+        raise salt.exceptions.ArgumentValueError("datacenter_name is required when looking up by cluster_name")
+
+    clusters = []
+    for cluster in get_mors_with_properties(
+            service_instance,
+            vim.ClusterComputeResource,
+            property_list=["name"]):
+        if cluster_name and cluster_name != cluster["name"]:
+            continue
+        if datacenter_name and datacenter_name != get_parent_of_type(cluster["object"], vim.Datacenter).name:
+            continue
+        clusters.append(cluster["object"])
+    return clusters
+
+
+def get_cluster(service_instance, datacenter_name, cluster_name):
+    """
+    Returns a vim.ClusterComputeResource managed object.
+
+    service_instance
+        The Service Instance Object from which to obtain datacenter.
+
+    datacenter_name
+        The datacenter name
+
+    cluster_name
+        The cluster name
+    """
+    if not cluster_name:
+        raise salt.exceptions.ArgumentValueError("Datacenter_name and cluster_name are required when looking up a cluster")
+    clusters = get_clusters(service_instance, datacenter_name, cluster_name)
+    if len(clusters) == 0:
+        return None
+    elif len(clusters) == 1:
+        return clusters[0]
+    else:
+        # this should be unreachable; if hit, it indicates a bug in this function
+        raise Exception()
```

Comment:
**question** would this read better like:

```
if len(clusters) > 1:
    raise WhoopsieDaisy('This should never have happened - please file a bug report at https://github.com/saltstack/salt-ext-modules-vmware/')
else:
    return clusters[0] if clusters else None
```

Reply:
When practicable I do like having the primary flow occur first; the equivalent would then be:

```
    if len(clusters) <= 1:
        return clusters[0] if clusters else None
    else:
        raise Bug('This should never happen - please file a bug report at https://github.com/saltstack/salt-ext-modules-vmware/')
```

I think it does read a little more clearly; I'm still not 100% on having a special "Bug" exception. I'll mull it over.

Reply:
I could also write it:

```
assert len(clusters) <= 1, "please file a bug report at https://github.com/saltstack/salt-ext-modules-vmware/"
return clusters[0] if clusters else None
```